### DISCUSSION TITLE
Constraint implementation

### DISCUSSION THREAD

**Discussion by nikitrian** [meta-pytorch/botorch#1747] [Q&A]
Hello,

My model works well without constraints. I have 9 dimensions and I decided to add a constraint in `optimize_acqf` such that
x[0]<=x[1], and I implemented it as:

```
        inequality_constraint = [
        (
        torch.tensor([0, 1, 2, 3, 4, 5, 6, 7, 8]),
        torch.tensor([1.0, -1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0], dtype=torch.float64),
        torch.tensor([0.0], dtype=torch.float64)
        )
            ]
```

 I get this warning message after iteration 3, 

> '[OptimizationWarning('Optimization failed within `scipy.optimize.minimize` with status 9.')] Trying again with a new set of initial conditions.'

After all the iterations are done, the results I am getting are as if the constraint does not exist and it is completely violated. Any ideas why?

Thanks

**Comment by saitcakmak** [comment]
Hi @nikitrian. Do you also see a warning like "Optimization failed on the second try, after generating a new set of initial conditions."? If so, this would mean the optimization failed on retry as well, in which case the solutions you get do not have any guarantees on satisfying the constraints. It might be useful to take a look at the output of scipy.minimize to see how that looks like (which is ran in `gen_candidates_scipy`).

**Reply by nikitrian** [reply]
Hi @saitcakmak, thanks for your reply. 
Yes, I see that too. How can I see the output of scipy.minimize? Do you think there might be something wrong with the constraint implementation?

**Reply by saitcakmak** [reply]
> Do you think there might be something wrong with the constraint implementation?

That's possible. I haven't worked with constraints much though, so I don't know what the failure modes are there.

> How can I see the output of scipy.minimize? 

You can either put some breakpoints in `gen_candidates_scipy` or mock one of the callsites where the result object is available. Something like this should work.
```
from unittest import mock
from botorch.generation.gen import _process_scipy_result
from botorch.optim import optimize_acqf

def mock_process_res(res, options):
    print(res)  # Do whatever you need here
    # You can also just put a breakpoint in here if you're running this in a notebook.
    _process_scipy_result(res, options)  # So that the original warnings are still there.

with mock.patch("botorch.generation.gen._process_scipy_result", wraps=mock_process_res):
    optimize_acqf(...)  # just as you'd normally.
```
This should print the `OptimizeResult` object returned by `scipy.minimize`.

**Reply by saitcakmak** [reply]
You can also add a `as <var_name>` after `mock.patch` to assign the mock object to a variable that you can later access. E.g., if you do `with mock.patch(...) as mock_res:`, you can later check `mock_res.call_args / call_args_list` to access the `res & options` that the function was called with for further investigation.
